### PR TITLE
Updated the intermine data source to use POST instead of GET requests

### PR DESCRIPTION
**Pull request by alancleary**
Switching to POST requests prevents URLs querystrings from being longer than the Intermine server supports.

I'm not sure if the POST requests (specifically the headers) are being made correctly. Therefore I'm opening a draft PR to emphasize the need for further review and testing.

### CODE REVIEW

**Comment by adf-ncgr** [comment]
@alancleary thanks for putting this up; I've been doing some testing and seem to be encountering some erratic behavior that is hard to pin down exactly what is going on as it seems to be something that is causing the proxy server to choke. I've actually had to restart apache several times on lupini. The only clue that seems possibly significant is an error in the proxy server logs:
`
server reached MaxRequestWorkers setting, consider raising the MaxRequestWorkers setting
`
as well as a bunch of these in the access log after the server is restarted:
```
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
172.20.130.2 - - [17/Oct/2024:23:17:35 -0600] "POST /glycinemine/service/query/results HTTP/1.1" 500 143 "-" "Bun/1.1.30"
```
which makes me suspect that something is causing more requests to be made than intended. 
It's a bit hard to describe the behavior so if you want to try to have a call tomorrow to look at it in real time let me know, I'm pretty free so far.

**Comment by alancleary** [comment]
@adf-ncgr Can you be more specific about which query is causing this behavior? In my testing I noticed that the query used to get data for the pangene lookup form seems to be generating a ton of requests, whereas the query to lookup pangenes seems to be well behaved.

**Comment by adf-ncgr** [comment]
Did you change anything about the way the queries for the form data are being made? It does seem odd that so many requests come in via graphql even before a gene list is provided, and I'm puzzled as to why this behavior would have changed though I haven't looked at your code closely (it didn't look like the changeover to POST had been a lot of lines of code though)

**Comment by alancleary** [comment]
I literally changed nothing besides the request type. I think the issue is that GraphQL resolves nested data types with additional queries. Given the poor representation of annotation and assembly versions in Intermine, the query I crafted to get the form data is pretty heavy handed. I'm guessing this previously wasn't an issue because we couldn't actually make requests that contained 100 genes!

**Comment by adf-ncgr** [comment]
Not sure that last hypothesis makes sense, given the fact that the form data queries are made before there's even a gene list whose length could influence anything? Maybe it was making lots of queries earlier and I just didn't notice it. Did the change from GET to POST affect all queries to to intermine? Maybe they are more heavyweight and this is now causing the issue.

**Comment by adf-ncgr** [comment]
Agreed that long term we ought to have a better way to populate the needed data for the forms. But I think supporting both GET and POST is probably a good idea. IMHO the workaround in this case is using POST due to URI length limits, it's not as if these queries are making updates to the backend. I would make GET the default and use POST only under special circumstances, since GET requests are cacheable and all that good idempotent stuff.

**Comment by alancleary** [comment]
OK, I pushed commits that make GET the default request type again while preserving support for POST requests via an alternate method. The pangene lookup component appears to be working as expected now.

**Comment by adf-ncgr** [comment]
trying to test this but having some difficulties with docker+bun build process. will let you know when I've been able to figure this out

**Comment by adf-ncgr** [comment]
Hmm, I'm still having issues with the proxy server getting into a hung state.I don't think this is the root of the problem, but FWIW I am noticing some GETs with long gene lists in their query strings, e.g.:
`172.20.130.2 - - [18/Oct/2024:11:02:32 -0600] "GET /glycinemine/service/query/results?query=%3Cquery+model%3D%27genomic%27+view%3D%27Gene.id+Gene.primaryIdentifier+Gene.description+Gene.symbol+Gene.name+Gene.assemblyVersion+Gene.annotationVersion+Gene.length+Gene.briefDescription+Gene.organism.taxonId+Gene.strain.identifier%27+sortOrder%3D%27Gene.primaryIdentifier%27+%3E%3Cconstraint+path%3D%27Gene.primaryIdentifier%27++op%3D%27ONE+OF%27%3E%3Cvalue%3Eglyma.Lee.gnm2.ann1.Gm_00001%3C%2Fvalue%3E%3Cvalue%3Eglyma.Lee.gnm2.ann1.Gm_00002...constraint%3E%3C%2Fquery%3E&format=jsoncount HTTP/1.1" 200 105 "-" "Bun/1.1.31"`
not sure why those aren't triggering the URI length error, but maybe a good idea to POSTify them anyway?

**Comment by alancleary** [comment]
@adf-ncgr At this point I think it would be best if we debugged this together in real time. Let's set something up via email.

**Comment by alancleary** [comment]
@adf-ncgr I pushed a couple commits that implement caching of PathQuery POST requests. For some reason I started having issues with not getting cache hits when I was expecting them during my testing, although it was working fine earlier today... Anyway, if you could take this for a spin to make sure it's behaving correctly that would be great!

Also note that I set the time to live to infinity since GraphQL is read-only and InterMine access patterns are write once read many. If the cache doesn't seem to be working maybe try changing this to an actual number.

**Comment by adf-ncgr** [comment]
thanks I will try spinning it momentarily.

**Comment by alancleary** [comment]
@adf-ncgr GET by default has been pushed. Let me know if you have any issues.

**Comment by adf-ncgr** [comment]
@alancleary overall seems better but still having some issues. I don't see any evidence that caching is actually taking place, e.g. when I reload the web components page the proxy server gets hit with a bunch of GETs from Bun (and the intermine tomcat server in turn gets hit by queries from the proxy). This may be due to the fact that the dev glycinemine seems to have reverted to sending back the cache-busting headers, probably due to getting the older version of the code when I had to redeploy the web app. I'll try to deal with that before complaining any further

**Comment by adf-ncgr** [comment]
Removing the headers again seemed to have no effect on the caching (ie reloading web component page continues to make many GET requests of the proxy and on to intermine). @alancleary when you were doing your testing how were you monitoring for cache hits? My method of monitoring the server logs seems crude but on the other hand hard to argue that traffic is getting there when the messages show up. I guess I'm just starting to wonder what evidence we have that the caching mechanism ever worked at all within the graph-ql server?

**Comment by adf-ncgr** [comment]
Another little bit of insight that may or may not be significant; when we were pair-debugging and the intermine server was in a bad state at one point I had a long stack trace which I later noticed had some info about the ultimate cause being due to insufficient db connections available:
`org.intermine.sql.DatabaseConnectionException: Unable to open database connection (there may not be enough available connections): {datasource.maxConnections=10, datasource.dataSourceName=db.production, datasource.user=intermine, ...`
I'm wondering how much of what we're seeing could potentially be attributable to overwhelming the db connection pool with concurrent requests. Not exactly sure what to do with this observation but thought it was at least worth mentioning.

**Comment by adf-ncgr** [comment]
(of course, if the cache were functioning properly then levels of concurrency on the db side would be less of an issue, so we should probably still focus on that)

**Comment by adf-ncgr** [comment]
new commit is definitely helping a lot with respect to caching. but the download of 100 genes can still overwhelm the proxy server, even in the context of a specific target genome, so I think we'll need to do one or more of the following:
- find a way to reduce the number of concurrent requests made in that context (my preference)
- find a way to make the proxy server/intermine more tolerant of many small requests (probably some apache config changes could help here)
- reduce the number of genes allowed in the supplied list (and hang our heads in shame- what organism besides a virus could make do with less than 100 genes?!)

open to other ideas, and willing to go with the third option in the short term as long we don't have to dumb it down too much.

**Comment by StevenCannon-USDA** [comment]
I can see that it would be nice to figure this out, but if the problem can be solved by lowering the allowed gene count (to, say, 50 or even 25), I don't think that's terrible. We'll be providing the full pangene table for each species; and if someone wants to do bulk translation (say, of 1000 genes or every gene), that's the way to go. We provide some instructions under "details" here: https://www.soybase.org/tools/translate/

**Comment by alancleary** [comment]
I'll see if I can do anything today about reducing the number of requests coming from GraphQL. If nothing comes of it let's just plan to reduce the number of genes allowed per query and ship!

**Comment by adf-ncgr** [comment]
OK, and I think we should start testing this against legumemine instead of glycinemine, since the "many requests from graphql to intermine even just to get the form dropdown data" will be most stressful for stress-testing in that context, and we don't have a workaround for that (although it didn't seem to be an issue previously, as far as I can recall).

**Comment by adf-ncgr** [comment]
Initial test of form load against legumemine went OK, then things appeared to go south afterwards but some apache log inspection suggests that some requests are coming from @alancleary doing testing against glycinemine, so I am going to back off for now so that we don't confuse each other with uncoordinated stressors....

**Comment by alancleary** [comment]
@adf-ncgr I pushed a commit that adds deduplication for POST request. Now GET and POST requests are only sent if there isn't a cached result AND the same request isn't already in progress.

**Comment by alancleary** [comment]
@adf-ncgr I've pushed commits that prevent HTTP requests for pangene pair subfield types when the query coming from the client only requests the identifier field of these types, i.e. data that's already been retrieved via our pangene pair outer join PathQuery! Give it a try with the pangene lookup component and let me know what you think.

**Comment by adf-ncgr** [comment]
OK, I will update and let you know how it goes. Hopefully you won't be testing any more today...

**Comment by adf-ncgr** [comment]
Initial impressions are very favorable! It doesn't seem to be breaking a sweat with 100 genes as input and the least restrictive target set I could make, yielding downloads of several thousand genes in seconds. I think it's probably a winner, and worth all the sweat and tears (next time I'll have to try to draw blood too). If you think you're up for making a release today I'd say go for it but I can also do some more testing and/or put a non-official release in place for the staged version of the lis site to talk to.

**Comment by alancleary** [comment]
I can make a release. One last question, though, now that all seems to be in order with POST requests, do we want to change that to be the default request type as not to get burned by URL overflows in the future?

**Comment by adf-ncgr** [comment]
I'm tempted to say "don't change a thing" but I guess I could go either way. One very minor point in favor of GETs whenever possible is that it's much easier to figure out what requests represent in the log files when everything you need is in the URL. That in itself isn't probably sufficient justification, so if you have very strong feelings about POST as default I'm happy to test it.

**Comment by alancleary** [comment]
Let's go with POST as default then. That's actually how I was testing all day, I just didn't commit the changes. Will commit it now...

**Comment by adf-ncgr** [comment]
seems to be working fine although I am still seeing GETs in my proxy server logs. will double check that I got the update but pretty sure I did...

**Comment by adf-ncgr** [comment]
last one was my bad but now that I've updated properly and seeing POSTs I also got this error 
```
    "errors": [
        {
            "message": "this.pathQueryPost is not a function. (In 'this.pathQueryPost(query, { page, pageSize })', 'this.pathQueryPost' is undefined)",
            "locations": [
                {
                    "line": 3,
                    "column": 9
                }
            ],
            "path": [
                "panGenePairs"
            ],
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR"
            }
```
did you forget to commit something or could this somehow be on my end again?

**Comment by adf-ncgr** [comment]
seems likely to have been an issue with my build process under docker compose. I think I'll be able to work around it but might not finish my testing before you are off so if you are confident and want to try to get it released go ahead.

**Comment by alancleary** [comment]
I'll just plan to roll a release first thing in the morning. If your testing turns up something unfavorable I'll switch back to GET as the default request type first. It should be done by the LIS meeting.

**Comment by adf-ncgr** [comment]
OK, thanks. I think I'm finally back in action although still puzzled as to why the build process stopped working at that point and what the actual issue was but I've at least worked around it for now. And it still seems to be working fine from the enduser perspective although it doesn't appear that the POSTs are now being cached as far as I can tell which if true makes it seem pretty mysterious as to what the issue was that started us down this long rabbithole in the first place!

**Comment by alancleary** [comment]
POSTs should definitely be cached at this point. I'll do some quick testing to investigate. I they are in fact not being cached then I'll revert to GET as the default request before rolling a release.

**Comment by adf-ncgr** [comment]
OK sounds good but FWIW even with the apparent non-caching of POSTs in my setup I haven't been able to break anything in my testing, so I can only surmise that the recent changes you made to optimize the number of requests made behind the scenes of the graphql must have really helped a lot!

**Comment by alancleary** [comment]
It turns out you still need include to `ttl` to force caching of POST requests. I just set it to the `222`, the same as the `max-age` value we're setting in every response header.

**Comment by adf-ncgr** [comment]
tested. post caching works. no other issues observed. ship it!

**Review by adf-ncgr — APPROVED** [review]
it's perfect!
